OpenRA.Mods.Common/Traits/Crushable.cs

```diff
@@ -67,7 +67,7 @@ bool ICrushable.CrushableBy(Actor self, Actor crusher, BitSet<CrushClass> crushC
 
 		LongBitSet<PlayerBitMask> ICrushable.CrushableBy(Actor self, BitSet<CrushClass> crushClasses)
 		{
-			if (IsTraitDisabled || !self.IsAtGroundLevel() || !Info.CrushClasses.Overlaps(crushClasses))
+			if (IsTraitDisabled || !Info.CrushClasses.Overlaps(crushClasses))
```

Comment:
This change needs to be mirrored in `CrushableInner` by removing:

```c#
// Only make actor crushable if it is on the ground.
if (!self.IsAtGroundLevel())
    return false;
```

Reply:
This is needed to units don't get crushed while in air

Reply:
The implementations of `bool ICrushable.CrushableBy` and `LongBitSet<PlayerBitMask> ICrushable.CrushableBy` have be in sync with each other. The latter only exists for the benefit of the Locomotor cache being able to cache the results.

If they're not in sync we still have the same underlying issue where the cache is storing incorrect information.

Reply:
It seems CrushableBy is for the locomotor, and CrushableInner is for the crush action

Reply:
`bool ICrushable.CrushableBy` is purely used to determine wether to notify crushable actors. I'll move the check for crate to `bool ICrushable.CrushableBy`

Reply:
The `bool` returning overload is used by several things to determine if an item is crushable - including the slow-path in Locomotor when it isn't able to use the cache.

Reply:
I moved the check to `AnyCrushables` on Mobile and Aircraft

Reply:
This also means we're are enforcing that all `ICrushable` items must be at ground level. This means `Mine` which didn't previously have such checks now also gains them.

`Aircraft.IsBlockedBy` and `Locomotor.IsBlockedBy` would also need checks.

And having added such checks to `Locomotor`, you need to add a check to `HPF` whose logic must remain in sync with `Locomotor`. And we're back to the original issue that caching `IsAtGroundLevel` is a problem because we won't be aware when it changes.

This approach is just moving some code around but doesn't solve the core of the issue about the inconsistent information. We've moved where the inconsistency is with some user visible results, but that just means bugs or crashes resulting from the inconsistency have moved elsewhere too.

I don't think this approach is fundamentally sound.

Reply:
> `Aircraft.IsBlockedBy` and `Locomotor.IsBlockedBy` would also need checks.

No, we want units to act as if units at the wrong ground level are crushable, the only thing we want to prevent is the crushing action. 

> I don't think this approach is fundamentally sound.

I think it's very sound. The only issue I see is that `IsAtGroundLevel` doesn't support custom movement layers. This prevented other layer crushing and it will continue to prevent it in the future

Reply:
> No, we want units to act as if units at the wrong ground level are crushable, the only thing we want to prevent is the crushing action.

Ah this makes sense with the cleaned up logic. The tangle before we preventing my ability to discern the logic. The actor is considered crushable regardless of height, but can only actually be crushed when it makes it to the ground. This logic I like.

This then sidesteps the locomotion blocking issue from my last comment. Because we aren't blocked by at all those worries go away.

>  The only issue I see is that IsAtGroundLevel doesn't support custom movement layers. This prevented other layer crushing and it will continue to prevent it in the future

Indeed, we'd probably need to introduce an ability for custom movement layers to reports if actors are on a "ground level" equivalent. An issue for another day.

----

This thread is resolved.